drivers: flash: spi_nor: optimizations and better Micron flash support

robhancocksed opened the pull request:
- Avoid providing redundant TX buffers to the SPI driver for the data portion of read commands where the TX data is not used. This is more efficient and also avoids passing uninitialized data to an external device.
- Add support for the flag status register used on some Micron flash devices, which improves error detection and in some cases must be read for the flash device to function properly.
- Add support for fast read commands using a dummy byte between the address and data phases, which often allows the flash device to operate at a higher SPI clock rate.

Note, there might be a compliance error raised, as I wasn't able to satisfy both clang-format and checkpatch with regard to how one line was formatted..

robhancocksed committed:
drivers: flash: spi_nor: optimize SPI buffer usage

This driver was providing SPI buffers for both TX and RX on the data
payload portion of read transfers, even though the TX buffer is not
meaningful in these cases. As well as being less efficient, this also
caused likely uninitialized data to be transferred to the device, which
is possibly problematic.

Update to not include the TX buffer for the read data payload SPI
transfer, so that the SPI driver can generate dummy TX data internally.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed committed:
drivers: flash: spi_nor: Added flag status register support

Some Micron (and possibly other) SPI NOR devices implement a flag status
register which provides more information on the success/failure of erase
and program operations. In addition to better error checking, some of
these devices actually don't function properly if the flag status
register is not read after a program operation (subsequent reads will
only return 0xFF bytes).

Add a device tree parameter to indicate that the flag status register is
supported. When specified, the flag status register will be used for
ready/error checks rather than the standard status register.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed commented:
Looks like the drivers.console.line_splitting.plus.some_option test is currently failing, for presumably unrelated reasons..

robhancocksed commented:
Ping.. any feedback?

de-nordic commented on drivers/flash/spi_nor.c:
```diff
@@ -381,19 +389,19 @@ static int spi_nor_access(const struct device *const dev,
-		}
-	};
+	bool has_dummy = (access & NOR_ACCESS_DUMMY_BYTE) != 0U;
+	uint8_t cmd_buf[6] = {opcode};
```
Needless variable name spawns unneeded changes in lines 419 and 423. Any style changes, renames got to other PR. Needless changes eclipse the subject of PR changes.

robhancocksed replied:
Removed variable rename.

de-nordic commented on drivers/flash/spi_nor.c:
```diff
@@ -381,19 +381,18 @@ static int spi_nor_access(const struct device *const dev,
-			.len = length
-		}
-	};
+	uint8_t cmd_buf[5] = {opcode};
```
This line gets modified in two separate commits.

robhancocksed replied:
Not sure how you suggest handling this? Please have a look at the current change set.

de-nordic commented on drivers/flash/spi_nor.c:
```diff
@@ -46,13 +46,15 @@ LOG_MODULE_REGISTER(spi_nor, CONFIG_FLASH_LOG_LEVEL);
-#define ANY_INST_HAS_WP_GPIOS DT_ANY_INST_HAS_PROP_STATUS_OKAY(wp_gpios)
-#define ANY_INST_HAS_HOLD_GPIOS DT_ANY_INST_HAS_PROP_STATUS_OKAY(hold_gpios)
-#define ANY_INST_USE_4B_ADDR_OPCODES DT_ANY_INST_HAS_BOOL_STATUS_OKAY(use_4b_addr_opcodes)
+#define ANY_INST_HAS_DPD                    DT_ANY_INST_HAS_BOOL_STATUS_OKAY(has_dpd)
+#define ANY_INST_HAS_T_EXIT_DPD             DT_ANY_INST_HAS_PROP_STATUS_OKAY(t_exit_dpd)
+#define ANY_INST_HAS_DPD_WAKEUP_SEQUENCE    DT_ANY_INST_HAS_PROP_STATUS_OKAY(dpd_wakeup_sequence)
+#define ANY_INST_HAS_RESET_GPIOS            DT_ANY_INST_HAS_PROP_STATUS_OKAY(reset_gpios)
+#define ANY_INST_HAS_WP_GPIOS               DT_ANY_INST_HAS_PROP_STATUS_OKAY(wp_gpios)
+#define ANY_INST_HAS_HOLD_GPIOS             DT_ANY_INST_HAS_PROP_STATUS_OKAY(hold_gpios)
+#define ANY_INST_USE_4B_ADDR_OPCODES        DT_ANY_INST_HAS_BOOL_STATUS_OKAY(use_4b_addr_opcodes)
```
use-fast-read should have one of these too, and if there is not device in a system setup that has the property this should evaluate to false.

robhancocksed replied:
This is a case where clang-format wanted the new code formatted differently, and so I ended up updating the surrounding block of code to match. I can leave the other lines alone if preferred.

de-nordic replied:
Don't run clang-format on pre-existing units.

robhancocksed replied:
Ended up using "clang-format off" here as well.

de-nordic replied:
> Ended up using "clang-format off" here as well.

![image](https://github.com/user-attachments/assets/4fb2a784-7af6-40b4-9f5c-75467078a0c3)

You do not have to run the [thing](https://docs.zephyrproject.org/latest/contribute/style/index.html#clang-format), it is optional. 

![image](https://github.com/user-attachments/assets/9b4c2c2d-c791-4d75-9359-be4e97354abd)

I value your commitment to make the thing look nicer, but please do not mix that effort with changes in logic.

robhancocksed replied:
I have removed the clang-format directives. The compliance check still reports failures locally, but appears to be passing in CI.

de-nordic commented on drivers/flash/spi_nor.c:
```diff
@@ -803,14 +871,31 @@ static int spi_nor_read(const struct device *dev, off_t addr, void *dest,
+	if (IS_ENABLED(ANY_INST_USE_4B_ADDR_OPCODES) && cfg->use_4b_addr_opcodes) {
 		if (addr > SPI_NOR_3B_ADDR_MAX) {
-			ret = spi_nor_cmd_addr_read_4b(dev, SPI_NOR_CMD_READ_4B, addr, dest, size);
+			if (cfg->use_fast_read) {
```
This should be conditional on existence of device that have the fast read. I you have a system that does not have fast read or you do not want is  (and do not enable it in dts), then this should not be compiled in.
Example here may be MCUboot, where users may prefer to choose smaller driver at a cost of slower boot.

robhancocksed replied:
Done.

robhancocksed commented on drivers/flash/spi_nor.c:
```diff
@@ -381,19 +389,19 @@ static int spi_nor_access(const struct device *const dev,
-	struct spi_buf spi_buf[2] = {
-		{
-			.buf = buf,
-			.len = 1,
-		},
-		{
-			.buf = data,
-			.len = length
-		}
-	};
+	bool has_dummy = (access & NOR_ACCESS_DUMMY_BYTE) != 0U;
+	uint8_t cmd_buf[6] = {opcode};
+	struct spi_buf spi_buf_tx[2] = {{
+						.buf = cmd_buf,
+						.len = 1,
+					},
+					{.buf = is_write ? data : NULL, .len = length}};
+	struct spi_buf spi_buf_rx[2] = {{
+						.buf = NULL,
+						.len = 1,
+					},
+					{.buf = data, .len = length}};
```
Agreed, though for some reason the clang-format formatting wanted it this way. Will see if the compliance checks will flag it if I use the multi-line style.

robhancocksed committed:
drivers: flash: spi_nor: Add fast read support

Most SPI NOR flash devices support a "fast read" command which uses
dummy bits between the address and the start of the data transfer. In
many cases, the maximum SPI clock speed of the device is lower for the
regular read command due to the limited time between the address and
data phases, so using the fast read command will remove this restriction
and allow for faster transfers.

Add a device tree flag to indicate that fast reads should be used for
the device.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed commented on drivers/flash/spi_nor.c:
```diff
@@ -474,21 +499,63 @@ static int spi_nor_access(const struct device *const dev,
-	ARG_UNUSED(poll_delay);
+	/* If flag status register is present, check it rather than the standard
+	 * status register since it allows better error detection (and some devices
+	 * that have it require it to be read after a program operation).
```
Reworded this.

robhancocksed commented on drivers/flash/spi_nor.c:
```diff
@@ -474,21 +499,63 @@ static int spi_nor_access(const struct device *const dev,
+		if (IS_ENABLED(CONFIG_SPI_NOR_SLEEP_WHILE_WAITING_UNTIL_READY)) {
+			/* Don't monopolise the CPU while waiting for ready */
+			k_sleep(poll_delay);
+		}
```
Removed.

de-nordic commented on drivers/flash/spi_nor.c:
```diff
@@ -45,6 +45,7 @@ LOG_MODULE_REGISTER(spi_nor, CONFIG_FLASH_LOG_LEVEL);
 #define SPI_NOR_MAX_ADDR_WIDTH 4
 #define SPI_NOR_3B_ADDR_MAX    0xFFFFFF
 
+/* clang-format off */
```
Please understand that now we have pointless "TODO" comments in code, somebody else has to address.

robhancocksed replied:
As noted, appears to be required in order to pass CI compliance checks. If there is another way to do this, please advise.

de-nordic replied:
Reported an issue that clang-format requests changes in lines that have not been touched: https://github.com/zephyrproject-rtos/zephyr/issues/89561, you can add there additional input if you want to.

kartben replied:
> As noted, appears to be required in order to pass CI compliance checks. If there is another way to do this, please advise.

Just ignore the clang **warning** reported by CI. It's not blocking and will not prevent the compliance check to pass.

robhancocksed replied:
I have removed the clang-format directives. The compliance check still reports failures locally, but appears to be passing in CI.

robhancocksed commented on drivers/flash/spi_nor.c:
```diff
@@ -496,8 +518,8 @@ static int spi_nor_wait_until_ready(const struct device *dev, k_timeout_t poll_d
 		/* If flag status register is present, check it rather than the standard
-		 * status register since it allows better error detection (and some devices
-		 * that have it require it to be read after a program operation).
+		 * status register since it allows better error detection. Also, some devices
```
Indeed, not sure how it ended up in the third one. Moved this change to the second commit.

robhancocksed commented on drivers/flash/spi_nor.c:
```diff
@@ -409,20 +428,26 @@ static int spi_nor_access(const struct device *const dev,
 	const struct spi_buf_set tx_set = {
-		.buffers = spi_buf,
+		.buffers = spi_buf_tx,
 		.count = (length != 0) ? 2 : 1,
```
Yes, it looks like this works and simplifies things a bit.
